### PR TITLE
Support AGP 7.2 and remove 4.2 back compat

### PR DESCRIPTION
**Pull request by khoben**
Upgrade AGP to 7.x with no backward compatibility with 4.x

The key difference is the changed package for `AndroidComponentsExtension` since AGP 7.2.0:
 - AGP 4.2 package is `com.android.build.api.extension`
 - AGP 7.2 package is `com.android.build.api.variant`

### CODE REVIEW

**Comment by SUPERCILEX** [comment]
Thanks for opening a PR! I'd actually prefer to break backwards compatibility since people who want to stick to 4.2 can keep using the current version of the plugin.

**Comment by khoben** [comment]
Ok, that sounds reasonable.  Should I close the pull request?

**Comment by SUPERCILEX** [comment]
If that's easier, sure. Otherwise, I'm fine with a new commit that switches to using VariantSelector.

**Comment by khoben** [comment]
> new commit that switches to using VariantSelector.

Did I understand correctly that I need to add a commit where some `VariantSelector` can be passed to `onVariants()`?

**Comment by SUPERCILEX** [comment]
Actually wait looking at this further I'm confused. Is the goal of this PR to support AGP 4.2? Based on the title, I thought the goal was to support AGP 7.2 but the types in that codepath don't seem to have changed so I'm assuming AGP 7.2 is **not** broken. TL;DR: Are you using AGP 4.2 and want support for it?

**Comment by khoben** [comment]
I used for a while with agp 7.0.4, but with the upgrade to 7.2.0 the plugin stopped working. So I added support for 7.2.0 without affecting other older versions.

**Comment by khoben** [comment]
Here is error log with `classpath 'com.android.tools.build:gradle:7.2.1'` and `classpath "com.supercilex.gradle:version-orchestrator:1.0.0-20211121.212215-8"`:
```
* What went wrong:
A problem occurred evaluating project ':app'.
> Failed to apply plugin 'com.supercilex.gradle.versions'.
   > Type com.android.build.api.extension.ApplicationAndroidComponentsExtension not present
```

**Comment by SUPERCILEX** [comment]
Ohhh, the imports are different: `com.android.build.api.extension.ApplicationAndroidComponentsExtension` vs `com.android.build.api.variant.ApplicationAndroidComponentsExtension`.

Ok, so what I'm saying is that this PR should just bump the AGP version to 7.0 and change the import.

**Comment by khoben** [comment]
> Ohhh, the imports are different: com.android.build.api.extension.ApplicationAndroidComponentsExtension vs com.android.build.api.variant.ApplicationAndroidComponentsExtension.

Exactly

> should just bump the AGP version to 7.0 and change the import.

but not backwards compatible with 4.x versions, right?

All right then, I'll change the title of the PR and remove the compat classes.

**Comment by SUPERCILEX** [comment]
> but not backwards compatible with 4.x versions, right?

Correct

File: plugin/src/main/kotlin/com/supercilex/gradle/versions/internal/Validation.kt [hunk 1]
```diff
@@ -20,11 +23,15 @@ internal fun validateRuntime() {
...
     }
 
-    check(agpVersion >= MIN_AGP_VERSION) {
-        """
-        |Version Orchestrator's minimum Android Gradle Plugin version is at least
-        |$MIN_AGP_VERSION and yours is $agpVersion. Find the latest version and upgrade
-        |instructions at https://developer.android.com/studio/releases/gradle-plugin.
-        """.trimMargin()
+    project.plugins.withType<AppPlugin> {
+        val agpVersion = project.extensions.findByType<ApplicationAndroidComponentsExtension>()?.pluginVersion
```
Comment (khoben) [inline plugin/src/main/kotlin/com/supercilex/gradle/versions/internal/Validation.kt:27]: In AGP 7.0 getting the version is not too straightforward, is this method acceptable?

Comment (SUPERCILEX) [reply plugin/src/main/kotlin/com/supercilex/gradle/versions/internal/Validation.kt:27]: Yup!

**Comment by SUPERCILEX** [comment]
Alrighty, thanks for the PR and apologies for all the confusion. :)

**Pull request merged**